### PR TITLE
expression: rewrite compare built-in functions

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This commit rewrite built-in compare-function signatures (LT, LT, EQ, NullEQ, GT, GE, NE) using new builtinFunc interface.

@shenli @coocood @hanfei1991 @zimulala @tiancaiamao @winoros @lamxTyler PTAL

### CODE REVIEW

File: expression/builtin_compare.go [hunk 1]
```diff
@@ -272,3 +275,233 @@ func (s *builtinCompareSig) eval(row []types.Datum) (d types.Datum, err error) {
 	}
 	return
 }
+
+func resOfCmp(res int, op opcode.Op) int64 {
```
Comment (shenli) [inline expression/builtin_compare.go]: Add comment for the function.

**Comment by bobotu** [comment]
Is there need a `builtinCompareTimeSig`? MySQL will convert constant to time type when another argument is a time type column. Would it be easier understand to have a `builtinCompareTimeSig` for this case? Or just using another data type when compare time column and constant?

And I have a little question about further implementation.:

- Will this new resolver convert constant's `Datum` to appropriate type, just like convert constant `20010101` to a Time Datum `2001-01-01`, and replace the old value with the converted value in `Constant.Value`? 

- And will this new resolver do constant folding while building Expression?   

These question are all about bugs like #2318. In this case, plan builder try to serialize the `ScalarFunction` and its args in `where` into `pb format` then send it to backend. If we cannot convert `Constant.Value` to correct type this bug cannot fix. And if we cannot do constant folding we will miss some constant which need convert, like `dt > 20010100+1`.

I cannot access the design doc now, so I post these question here.

**Comment by coocood** [comment]
@bobotu 
Good suggestion!
A compareTimeSig makes the code easier to understand.
We can do constant folding while building Expression by adding a flag attribute to the expression.

**Comment by XuHuaiyu** [comment]
@bobotu Sorry for a late reply, I've sent the doc to you, please check your Gmail inbox.

**Comment by XuHuaiyu** [comment]
@coocood For `date/time/datetime column cmp

**Comment by bobotu** [comment]
Got it! I'll take a closely look. @XuHuaiyu

File: expression/builtin_compare.go [hunk 2]
```diff
@@ -17,9 +17,12 @@ import (
 	"sort"
 
 	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
 	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/mysql"
 	"github.com/pingcap/tidb/parser/opcode"
 	"github.com/pingcap/tidb/util/types"
+	"math"
```
Comment (tiancaiamao) [inline expression/builtin_compare.go]: move to line 18

File: expression/builtin_compare.go [hunk 3]
```diff
@@ -272,3 +275,234 @@ func (s *builtinCompareSig) eval(row []types.Datum) (d types.Datum, err error) {
...
+			if isKindNull0 && isKindNull1 {
+				return oneI64, false, nil
+			}
+			return zeroI64, false, nil
+		}
+		return zeroI64, true, nil
+	}
+	isUnsigned0, isUnsigned1 := mysql.HasUnsignedFlag(s.args[0].GetType().Flag), mysql.HasUnsignedFlag(s.args[1].GetType().Flag)
+	var res int
+	if !isUnsigned0 && !isUnsigned1 {
```
Comment (tiancaiamao) [inline expression/builtin_compare.go]: maybe change to `switch case` make this code more clear?

**Comment by XuHuaiyu** [comment]
@tiancaiamao PTAL

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**